src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -1037,31 +1066,39 @@ private void WriteCrcAndSizesInLocalHeader(bool zip64HeaderUsed)
             // and setting the version to Zip64 to indicate that descriptor contains 64-bit values
             if (pretendStreaming)
             {
+                int relativeVersionToExtractLocation = ZipLocalFileHeader.FieldLocations.VersionNeededToExtract - ZipLocalFileHeader.FieldLocations.VersionNeededToExtract;
+                int relativeGeneralPurposeBitFlagsLocation = ZipLocalFileHeader.FieldLocations.GeneralPurposeBitFlags - ZipLocalFileHeader.FieldLocations.VersionNeededToExtract;
+
                 VersionToExtractAtLeast(ZipVersionNeededValues.Zip64);
                 _generalPurposeBitFlag |= BitFlagValues.DataDescriptor;
 
-                _archive.ArchiveStream.Seek(_offsetOfLocalHeader + ZipLocalFileHeader.OffsetToVersionFromHeaderStart,
+                _archive.ArchiveStream.Seek(_offsetOfLocalHeader + ZipLocalFileHeader.FieldLocations.VersionNeededToExtract,
                                             SeekOrigin.Begin);
-                writer.Write((ushort)_versionToExtract);
-                writer.Write((ushort)_generalPurposeBitFlag);
+                BinaryPrimitives.WriteUInt16LittleEndian(writeBuffer[relativeVersionToExtractLocation..], (ushort)_versionToExtract);
+                BinaryPrimitives.WriteUInt16LittleEndian(writeBuffer[relativeGeneralPurposeBitFlagsLocation..], (ushort)_generalPurposeBitFlag);
+
+                _archive.ArchiveStream.Write(writeBuffer[..MetadataBufferLength]);
             }
 
             // next step is fill out the 32-bit size values in the normal header. we can't assume that
             // they are correct. we also write the CRC
-            _archive.ArchiveStream.Seek(_offsetOfLocalHeader + ZipLocalFileHeader.OffsetToCrcFromHeaderStart,
+            _archive.ArchiveStream.Seek(_offsetOfLocalHeader + ZipLocalFileHeader.FieldLocations.Crc32,
                                             SeekOrigin.Begin);
             if (!pretendStreaming)
             {
-                writer.Write(_crc32);
-                writer.Write(compressedSizeTruncated);
-                writer.Write(uncompressedSizeTruncated);
+                int relativeCrc32Location = ZipLocalFileHeader.FieldLocations.Crc32 - ZipLocalFileHeader.FieldLocations.Crc32;
+                int relativeCompressedSizeLocation = ZipLocalFileHeader.FieldLocations.CompressedSize - ZipLocalFileHeader.FieldLocations.Crc32;
+                int relativeUncompressedSizeLocation = ZipLocalFileHeader.FieldLocations.UncompressedSize - ZipLocalFileHeader.FieldLocations.Crc32;
```

Comment:
I think these could be `const`, like the variables declared at the top of the method.

Reply:
They can't at the moment - many of the FieldLocations members are `readonly static` variables. I can't make those `const` because they all ultimately reference to FieldLengths.Signature. FieldLengths.Signature, in turn, references the length of SignatureConstantBytes - and that's not a compile-time constant.

It'd be good to make this more consistent, but I'd need to change `FieldLengths.Signature` to be a magic number of 4 and we'd lose a bit of clarity because we're no longer directly referencing `SignatureConstantBytes.Length`. I could do that and add a comment though?

From running this through sharplab, they both shake out as JIT-time constants - it's just a style choice between clarity and consistency...

Reply:
Ah got it. Nevermind then.